### PR TITLE
Clean logs and job results

### PR DESCRIPTION
**Pull request by Martchus**
using `keep...` columns of job group

### CODE REVIEW

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 1]
```diff
@@ -160,6 +160,10 @@ __PACKAGE__->add_columns(
         data_type   => 'timestamp',
         is_nullable => 1,
     },
+    logs_present => {
+        data_type     => 'integer',
```
Comment (coolo) [inline lib/OpenQA/Schema/Result/Jobs.pm]: boolean sounds more like it, no?

File: lib/OpenQA/Schema/Result/JobGroups.pm [hunk 2]
```diff
@@ -156,28 +156,54 @@ sub important_builds {
...
 }
 
-# list all jobs that are expired
-sub expired_jobs {
-    my ($self) = @_;
+sub _find_expired_jobs {
+    my ($self, $keep_in_days, $keep_important_in_days) = @_;
 
     my @ors;
     my $important_builds = $self->important_builds;
```
Comment (coolo) [inline lib/OpenQA/Schema/Result/JobGroups.pm]: as this is an quite expensive call, please cache it between results and log cleanup

File: lib/OpenQA/Schema/Result/JobGroups.pm [hunk 3]
```diff
@@ -156,28 +156,54 @@ sub important_builds {
...
+    if ($keep_important_in_days) {
         # expired jobs in important builds
-        my $timecond
-          = {'<' => time2str('%Y-%m-%d %H:%M:%S', time - 24 * 3600 * $self->keep_important_results_in_days, 'UTC')};
+        my $timecond = {'<' => time2str('%Y-%m-%d %H:%M:%S', time - 24 * 3600 * $keep_important_in_days, 'UTC')};
         push(@ors, {BUILD => {-in => $important_builds}, t_finished => $timecond});
     }
-    my $jobs = $self->jobs->search({-or => \@ors}, {order_by => qw/id/});
-    return [$jobs->all];
+    return $self->jobs->search({-or => \@ors}, {order_by => qw/id/});
```
Comment (coolo) [inline lib/OpenQA/Schema/Result/JobGroups.pm]: as you're touching this, please use qw as function

File: lib/OpenQA/WebAPI.pm [hunk 4]
```diff
@@ -439,16 +439,18 @@ sub startup {
...
-    $self->gru->add_task(limit_assets       => \&OpenQA::Schema::Result::Assets::limit_assets);
-    $self->gru->add_task(download_asset     => \&OpenQA::Schema::Result::Assets::download_asset);
-    $self->gru->add_task(scan_old_jobs      => \&OpenQA::Schema::Result::Needles::scan_old_jobs);
-    $self->gru->add_task(scan_needles       => \&OpenQA::Schema::Result::Needles::scan_needles);
-    $self->gru->add_task(migrate_images     => \&OpenQA::Schema::Result::JobModules::migrate_images);
-    $self->gru->add_task(relink_testresults => \&OpenQA::Schema::Result::JobModules::relink_testresults);
-    $self->gru->add_task(rm_compat_symlinks => \&OpenQA::Schema::Result::JobModules::rm_compat_symlinks);
-    $self->gru->add_task(scan_images        => \&OpenQA::Schema::Result::Screenshots::scan_images);
-    $self->gru->add_task(scan_images_links  => \&OpenQA::Schema::Result::Screenshots::scan_images_links);
+    # reduce_result is obsolete (replaced by limit_logs)
```
Comment (coolo) [inline lib/OpenQA/WebAPI.pm]: outdated comment :)

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 5]
```diff
@@ -912,28 +914,29 @@ sub part_of_important_build {
...
+sub delete_logs {
+    my ($self) = @_;
 
-    if ($args->{jobid}) {
-        my $job = $app->db->resultset('Jobs')->find({id => $args->{jobid}});
-        if ($job && $job->part_of_important_build) {
-            $app->log->debug('Job ' . $job->id . ' is part of important build, skip cleanup');
-            return;
-        }
+    if ($self->part_of_important_build) {
```
Comment (coolo) [inline lib/OpenQA/Schema/Result/Jobs.pm]: where else are you calling this function? As this is (again) a very expensive call I would just leave that decision to the caller

**Comment by coolo** [comment]
What level of  documentation are you talking about?

**Comment by okurz** [comment]
> What level of documentation are you talking about?

high level documentation for end users

**Comment by coolo** [comment]
ah. not sure why you want this high level documentation as part of this PR though

**Comment by okurz** [comment]
One has to start somewhere

**Comment by coolo** [comment]
#1032 would have been a better PR to start requiring that.

**Comment by coolo** [comment]
the test suite failure looks very interesting

**Comment by coolo** [comment]
To fix deploy.t:

```
--- a/dbicdh/_common/upgrade/47-48/03-move-resultdirs.pl
+++ b/dbicdh/_common/upgrade/47-48/03-move-resultdirs.pl
@@ -25,7 +25,7 @@ use OpenQA::Utils;
 sub {
     my ($schema) = @_;
 
-    my $jobs = $schema->resultset('Jobs')->search({});
+    my $jobs = $schema->resultset('Jobs')->search({}, {select => [qw(id result_dir)]});
 
     while (my $job = $jobs->next) {
         next unless $job->result_dir;
```

Using models in migrations is unfortunately very touchy

**Comment by okurz** [comment]
> #1032 would have been a better PR to start requiring that.

I am thinking of a mere two lines in GettingStarted.asciidoc at least. That should be feasible.

**Comment by Martchus** [comment]
@coolo Thanks for the fix, I included it.

> I am thinking of a mere two lines in GettingStarted.asciidoc at least. That should be feasible.

Good idea, but the concept of job groups isn't explained in the documentation at all. But we need to explain what a job group is first because settings for this cleanup features are configured on job group level and might inherit settings from parent groups.

Hence I think this goes out of the scope of this pull request. But I can work on it of course.

**Comment by okurz** [comment]
still, documentation is missing. @Martchus is practically the most active contributor to openQA lately so I think he will also have the most capacity to extend the documentation accordingly. As suggested, could be just 2 lines to an existing document but I guess it shouldn't harm to add another file and reference that accordingly.

**Comment by foursixnine** [comment]
@Martchus You can always create one new file explaining what your feature does, i can add the jobgroups description later in the Getting Started Guide and link it accordingly to this part... Don't feel blocked because the jobgroups docs are missing

**Comment by okurz** [comment]
oh, and also: The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a fate entry, the choices you made, etc. Also see http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages.

**Comment by coolo** [comment]
I'm with @Martchus here - this is out of scope

**Pull request merged**